File: php/tests/RunnerTest.php
#!/usr/bin/env php
<?php

require_once __DIR__ . '/../src/CommandSpec.php';
require_once __DIR__ . '/../src/CompletedRun.php';
require_once __DIR__ . '/../src/Runner.php';
require_once __DIR__ . '/../src/build_prompt_spec.php';

use Call\Coding\Clis\CommandSpec;
use Call\Coding\Clis\CompletedRun;
use Call\Coding\Clis\Runner;
use function Call\Coding\Clis\build_prompt_spec;

$passed = 0;
$failed = 0;

function assert_test(string $name, bool $condition): void
{
    global $passed, $failed;
    if ($condition) {
        $passed++;
        echo "ok - $name\n";
    } else {
        $failed++;
        echo "not ok - $name\n";
    }
}

$spec = build_prompt_spec("hello");
assert_test('valid prompt argv', $spec->argv === ['opencode', 'run', 'hello']);

try {
    build_prompt_spec("");
    assert_test('empty prompt throws', false);
} catch (\InvalidArgumentException $e) {
    assert_test('empty prompt throws', str_contains($e->getMessage(), 'empty'));
}

try {
    build_prompt_spec("   ");
    assert_test('whitespace-only prompt throws', false);
} catch (\InvalidArgumentException $e) {
    assert_test('whitespace-only prompt throws', str_contains($e->getMessage(), 'empty'));
}

$trimmed = build_prompt_spec("  foo  ");
assert_test('whitespace trimmed', $trimmed->argv === ['opencode', 'run', 'foo']);

try {
    build_prompt_spec(null);
    assert_test('null prompt throws', false);
} catch (\InvalidArgumentException $e) {
    assert_test('null prompt throws', str_contains($e->getMessage(), 'empty'));
}

$mockRunner = new Runner(function (CommandSpec $spec): CompletedRun {
    return new CompletedRun(['echo', 'hello'], 0, "hello\n", '');
});

$result = $mockRunner->run(build_prompt_spec("test"));
assert_test('mock exit code', $result->exit_code === 0);
assert_test('mock stdout', $result->stdout === "hello\n");
assert_test('mock stderr', $result->stderr === '');

$realRunner = new Runner();
$badSpec = new CommandSpec(['/nonexistent_binary_xyz']);
$badResult = $realRunner->run($badSpec);
assert_test('startup failure format', str_starts_with($badResult->stderr, "failed to start /nonexistent_binary_xyz:"));
assert_test('startup failure exit code', $badResult->exit_code === 1);

$stubScript = <<<'SH'
#!/bin/sh
if [ "$1" != "run" ]; then exit 9; fi
shift
printf 'opencode run %s\n' "$1"
SH;

$tmpFile = tempnam(sys_get_temp_dir(), 'ccc_test_');
file_put_contents($tmpFile, $stubScript);
chmod($tmpFile, 0755);

function run_ccc(string $prompt, string $stubPath): array
{
    $descriptors = [
        0 => ['pipe', 'r'],
        1 => ['pipe', 'w'],
        2 => ['pipe', 'w'],
    ];
    $env = array_merge(getenv(), ['CCC_REAL_OPENCODE' => $stubPath]);
    $proc = proc_open(
        [PHP_BINARY, __DIR__ . '/../bin/ccc', $prompt],
        $descriptors,
        $pipes,
        null,
        $env
    );
    fclose($pipes[0]);
    $stdout = stream_get_contents($pipes[1]);
    $stderr = stream_get_contents($pipes[2]);
    fclose($pipes[1]);
    fclose($pipes[2]);
    $exitCode = proc_close($proc);
    return [$stdout ?? '', $stderr ?? '', $exitCode];
}

list($out, $err, $rc) = run_ccc("Fix the failing tests", $tmpFile);
assert_test('ccc happy path exit code', $rc === 0);
assert_test('ccc happy path stdout', $out === "opencode run Fix the failing tests\n");

list($out, $err, $rc) = run_ccc("", $tmpFile);
assert_test('ccc empty prompt rejected', $rc !== 0);

unlink($tmpFile);

echo "\n";
echo "Results: $passed passed, $failed failed\n";
exit($failed > 0 ? 1 : 0);

File: zig/tests/runner_test.zig
const std = @import("std");
const runner = @import("runner");
const prompt_spec = @import("prompt_spec");

var stream_events: std.ArrayList(struct { []const u8, []const u8 }) = undefined;
var stream_alloc: std.mem.Allocator = undefined;

fn onStreamEvent(event_type: []const u8, data: []const u8) void {
    stream_events.append(stream_alloc, .{ event_type, data }) catch {};
}

test "buildPromptSpec valid prompt" {
    const allocator = std.testing.allocator;
    const spec = try prompt_spec.buildPromptSpec(allocator, "hello");
    defer prompt_spec.freePromptSpec(allocator, spec);

    try std.testing.expectEqualStrings("opencode", spec.argv[0]);
    try std.testing.expectEqualStrings("run", spec.argv[1]);
    try std.testing.expectEqualStrings("hello", spec.argv[2]);
}

test "buildPromptSpec empty prompt" {
    const allocator = std.testing.allocator;
    const result = prompt_spec.buildPromptSpec(allocator, "");
    try std.testing.expectError(error.EmptyPrompt, result);
}

test "buildPromptSpec whitespace-only prompt" {
    const allocator = std.testing.allocator;
    const result = prompt_spec.buildPromptSpec(allocator, "   ");
    try std.testing.expectError(error.EmptyPrompt, result);
}

test "buildPromptSpec trims whitespace" {
    const allocator = std.testing.allocator;
    const spec = try prompt_spec.buildPromptSpec(allocator, "  foo  ");
    defer prompt_spec.freePromptSpec(allocator, spec);

    try std.testing.expectEqualStrings("foo", spec.argv[2]);
}

test "Runner run echo" {
    const allocator = std.testing.allocator;

    const spec = runner.CommandSpec{
        .argv = &.{ "echo", "hello" },
    };

    const result = try runner.Runner.run(allocator, spec);
    defer result.deinit();

    try std.testing.expectEqual(@as(u8, 0), result.exit_code);
    try std.testing.expect(std.mem.indexOf(u8, result.stdout_data, "hello") != null);
    try std.testing.expectEqualStrings("", result.stderr_data);
}

test "Runner run nonexistent binary" {
    const allocator = std.testing.allocator;

    const spec = runner.CommandSpec{
        .argv = &.{"/nonexistent_binary_xyz"},
    };

    const result = try runner.Runner.run(allocator, spec);
    defer result.deinit();

    try std.testing.expectEqual(@as(u8, 1), result.exit_code);
    try std.testing.expect(std.mem.startsWith(u8, result.stderr_data, "failed to start /nonexistent_binary_xyz:"));
    try std.testing.expectEqualStrings("", result.stdout_data);
}

test "Runner run exit code forwarding" {
    const allocator = std.testing.allocator;

    const spec = runner.CommandSpec{
        .argv = &.{ "sh", "-c", "exit 42" },
    };

    const result = try runner.Runner.run(allocator, spec);
    defer result.deinit();

    try std.testing.expectEqual(@as(u8, 42), result.exit_code);
}

test "Runner stream fires callbacks" {
    const allocator = std.testing.allocator;

    var events: std.ArrayList(struct { []const u8, []const u8 }) = .empty;
    defer events.deinit(allocator);

    stream_events = events;
    stream_alloc = allocator;

    const spec = runner.CommandSpec{
        .argv = &.{ "echo", "stream_test" },
    };

    const result = try runner.Runner.stream(allocator, spec, onStreamEvent);
    defer result.deinit();

    events = stream_events;

    try std.testing.expectEqual(@as(usize, 1), events.items.len);
    try std.testing.expectEqualStrings("stdout", events.items[0][0]);
    try std.testing.expect(std.mem.indexOf(u8, events.items[0][1], "stream_test") != null);
}
